src/base/Optional.h

```diff
@@ -122,6 +122,53 @@ class Optional
     bool hasValue_ = false;
 };
 
+/// Specialization to make Optional<bool> trivially-copyable. XXX: Keep this
+/// temporary (until we switch to C++17 std::optional) hack in sync with the
+/// generic Optional above, copying generic methods where possible.
+template <>
+class Optional<bool>
+{
+public:
+    using Value = bool;
+
+    constexpr Optional() noexcept = default;
+    constexpr explicit Optional(const Value &v): value_(v), hasValue_(true) {}
+
+    constexpr explicit operator bool() const noexcept { return hasValue_; }
+    constexpr bool has_value() const noexcept { return hasValue_; }
+
+    const Value &value() const &
+    {
+        if (!hasValue_)
+            throw BadOptionalAccess();
+        return value_;
+    }
+
+    template <class Other>
+    constexpr Value value_or(Other &&defaultValue) const &
+    {
+        return hasValue_ ? value_ : static_cast<Value>(std::forward<Other>(defaultValue));
+    }
+
+    template <class Other = Value>
+    Optional &operator =(Other &&otherValue)
+    {
+        value_ = std::forward<Other>(otherValue);
+        hasValue_ = true;
```

Comment:
> This looks like invalid move semantics

It does look invalid, but it is the correct semantics in this case. See generic Optional docs for this method:

```C++
Optional(Optional<Value> &&other): Optional()
{
    if (other.hasValue_) {
        *this = std::move(other.value_);
        // no other.reset() per std::optional move semantics
    }
}
```

From https://en.cppreference.com/w/cpp/utility/optional/optional:

> 3) Move constructor: If other contains a value, initializes the contained value as if [direct-initializing](https://en.cppreference.com/w/cpp/language/direct_initialization) (but not direct-list-initializing) an object of type T with the expression std::move(*other) and does not make other empty: a moved-from optional still contains a value, but the value itself is moved from.

Reply:
:exclamation: My earlier [response](https://github.com/squid-cache/squid/pull/1188/files#r1035003281) more-or-less applies to the change request [suggestion](https://github.com/squid-cache/squid/pull/1188/files#r1034914264), but not to the PR code in question: This change request suggestion implied that the change request is about a standard move assignment operator. My earlier response discussed a standard move _constructor_ instead (i.e. item 3 on cppreference as quoted in my earlier response). Sorry. However, the standard move assignment follows exactly the same invalid-looking logic (see item 3 on the corresponding [cppreference](https://en.cppreference.com/w/cpp/utility/optional/operator%3D) page):

> 3). Note that a moved-from optional still contains a value. 

The PR _code_ in question is _not_ a standard move assignment operator (i.e. not item 3)! This operator works on a value object rather than on an Optional<Value> object. It is actually a value conversion/moving assignment operator (for the lack of a better term). Here, we are creating an Optional<Value> object from a Value-compatible value object. That is why hasValue_ is set to true unconditionally -- we _were_ given a value. This operator is described in item 4 on [cppreference](https://en.cppreference.com/w/cpp/utility/optional/operator%3D) page for `std::optional::operator =`.

In summary, there are no known problems in this PR code, but this thread can be used as one more illustration of how complex and tricky std::optional details can be!